File: lib/features/development/data/services/gitignore_service.dart
/// Service for parsing and checking .gitignore files
library;

import 'dart:io';
import 'package:path/path.dart' as path;

/// Service for handling .gitignore logic
class GitIgnoreService {
  final Map<String, List<String>> _cache = {};

  /// Check if a file/directory path is gitignored
  ///
  /// [filePath] - Absolute path to check
  /// [repositoryPath] - Root path of the git repository
  Future<bool> isIgnored(String filePath, String repositoryPath) async {
    final patterns = await _getIgnorePatterns(repositoryPath);
    if (patterns.isEmpty) return false;

    // Get relative path from repository root
    final relativePath = path.relative(filePath, from: repositoryPath);

    // Normalize path separators for matching
    final normalizedPath = relativePath.replaceAll('\\', '/');

    for (final pattern in patterns) {
      if (_matchesPattern(normalizedPath, pattern)) {
        return true;
      }
    }

    return false;
  }

  /// Get all ignore patterns from .gitignore file
  Future<List<String>> _getIgnorePatterns(String repositoryPath) async {
    // Check cache first
    if (_cache.containsKey(repositoryPath)) {
      return _cache[repositoryPath]!;
    }

    final patterns = <String>[];
    final gitignoreFile = File(path.join(repositoryPath, '.gitignore'));

    if (await gitignoreFile.exists()) {
      final lines = await gitignoreFile.readAsLines();

      for (final line in lines) {
        final trimmed = line.trim();

        // Skip empty lines and comments
        if (trimmed.isEmpty || trimmed.startsWith('#')) {
          continue;
        }

        patterns.add(trimmed);
      }
    }

    // Cache the patterns
    _cache[repositoryPath] = patterns;

    return patterns;
  }

  /// Match a path against a gitignore pattern
  bool _matchesPattern(String filePath, String pattern) {
    // Handle negation patterns (!)
    if (pattern.startsWith('!')) {
      return false; // Negation patterns are complex, skip for now
    }

    // Store original pattern for directory check
    final isDirectoryPattern = pattern.endsWith('/');

    // Remove leading slash
    if (pattern.startsWith('/')) {
      pattern = pattern.substring(1);
    }

    // Remove trailing slash for directory patterns
    if (isDirectoryPattern) {
      pattern = pattern.substring(0, pattern.length - 1);
    }

    // Build regex pattern - need to handle ** and * differently
    // Convert pattern to regex
    var regexPattern = pattern
        .replaceAll('.', '\\.')
        .replaceAllMapped(RegExp(r'\*\*'), (m) => '<<DOUBLESTAR>>')
        .replaceAll('*', '[^/]*')
        .replaceAll('<<DOUBLESTAR>>', '.*')
        .replaceAll('?', '[^/]');

    // Handle different pattern types
    if (isDirectoryPattern) {
      // Directory pattern - match directory name anywhere in path
      if (!pattern.contains('/')) {
        regexPattern = '(^|/)$regexPattern(/.*)?';
      } else {
        regexPattern = '^$regexPattern(/.*)?';
      }
    } else if (!pattern.contains('/')) {
      // Pattern without / matches filename anywhere in tree
      regexPattern = '(^|/)$regexPattern\$';
    } else {
      // Pattern with / matches from root
      regexPattern = '^$regexPattern\$';
    }

    try {
      final regex = RegExp(regexPattern);
      return regex.hasMatch(filePath);
    } catch (e) {
      // Invalid regex, skip this pattern
      return false;
    }
  }

  /// Clear the cache (useful when .gitignore changes)
  void clearCache([String? repositoryPath]) {
    if (repositoryPath != null) {
      _cache.remove(repositoryPath);
    } else {
      _cache.clear();
    }
  }
}

File: src/main/services/gitignore.service.ts
import ignore, { Ignore } from 'ignore';
import * as fs from 'fs';
import * as path from 'path';

/**
 * GitIgnore Service
 *
 * Parses and evaluates .gitignore patterns for file filtering
 */
export class GitIgnoreService {
  private ignoreCache: Map<string, Ignore>;

  constructor() {
    this.ignoreCache = new Map();
  }

  /**
   * Load .gitignore file and create ignore instance
   */
  private async loadGitIgnore(repoPath: string): Promise<Ignore> {
    const gitignorePath = path.join(repoPath, '.gitignore');
    const ig = ignore();

    // Always ignore .git directory
    ig.add('.git');

    try {
      if (fs.existsSync(gitignorePath)) {
        const content = await fs.promises.readFile(gitignorePath, 'utf-8');
        const patterns = content
          .split('\n')
          .map((line) => line.trim())
          .filter((line) => line && !line.startsWith('#')); // Remove empty lines and comments

        ig.add(patterns);
      }
    } catch {
      // .gitignore loading is best-effort
    }

    return ig;
  }

  /**
   * Get or create ignore instance for a repository
   */
  private async getIgnore(repoPath: string): Promise<Ignore> {
    if (!this.ignoreCache.has(repoPath)) {
      const ig = await this.loadGitIgnore(repoPath);
      this.ignoreCache.set(repoPath, ig);
    }

    return this.ignoreCache.get(repoPath)!;
  }

  /**
   * Check if a file path is ignored by .gitignore
   */
  async isIgnored(repoPath: string, filePath: string): Promise<boolean> {
    try {
      const ig = await this.getIgnore(repoPath);

      // Convert absolute path to relative path from repo root
      const relativePath = path.relative(repoPath, filePath);

      // Normalize path separators for ignore library (always use forward slashes)
      const normalizedPath = relativePath.split(path.sep).join('/');

      return ig.ignores(normalizedPath);
    } catch {
      return false;
    }
  }

  /**
   * Get all gitignore patterns for a repository
   */
  async getPatterns(repoPath: string): Promise<string[]> {
    const gitignorePath = path.join(repoPath, '.gitignore');

    try {
      if (fs.existsSync(gitignorePath)) {
        const content = await fs.promises.readFile(gitignorePath, 'utf-8');
        return content
          .split('\n')
          .map((line) => line.trim())
          .filter((line) => line && !line.startsWith('#'));
      }
      return [];
    } catch {
      return [];
    }
  }

  /**
   * Filter an array of file paths, removing ignored files
   */
  async filterIgnored(repoPath: string, filePaths: string[]): Promise<string[]> {
    const ig = await this.getIgnore(repoPath);

    return filePaths.filter((filePath) => {
      const relativePath = path.relative(repoPath, filePath);
      const normalizedPath = relativePath.split(path.sep).join('/');
      return !ig.ignores(normalizedPath);
    });
  }

  /**
   * Reload gitignore patterns for a repository
   */
  async reload(repoPath: string): Promise<void> {
    this.ignoreCache.delete(repoPath);
    await this.loadGitIgnore(repoPath);
  }

  /**
   * Clear all cached gitignore patterns
   */
  clearCache(): void {
    this.ignoreCache.clear();
  }

  /**
   * Check if .gitignore file exists
   */
  async hasGitIgnore(repoPath: string): Promise<boolean> {
    const gitignorePath = path.join(repoPath, '.gitignore');
    try {
      await fs.promises.access(gitignorePath);
      return true;
    } catch {
      return false;
    }
  }
}

// Export singleton instance
export const gitIgnoreService = new GitIgnoreService();
